src/Nova-GPU-Driver.md

```diff
@@ -1,62 +1,110 @@
+
+Besides Rust's built-in ownership and lifetime model, its powerful type system
+allows us to avoid a large portion of a whole class of bugs (i.e. memory safety
+bugs).
 
-In the source tree, the driver lives in `drivers/gpu/drm/nova`.
+Additionally, the same features allow us to model APIs in a way that also
+certain logic errors can be caught at compile time already.
 
+Those are errors GPU drivers suffer from significantly, because GPUs are, by
+definition, asynchronous in regards to the CPU and can handle a great many jobs
+(i.e., memory objects) simultaneously. Jobs' status can be changed at different
+places in the code base at different points in time (through work items,
+interrupt handlers, userspace calls, ...).
 
-## Status
+In short, GPU drivers are expected to profit the most from the promised memory
+safety.
 
-Currently, Nova is just a stub driver intended to lift the bindings necessary
-for a real GPU driver into the (mainline) kernel.
+Since Nova is a freshly written new driver, it is an opportunity to try to
+leverage the advantages of Rust and obtain a more reliable, maintainable driver.
 
-Currently, those efforts are mostly focused on getting bindings for PCI, DRM
-and the Device (driver) model upstream.
 
-It can be expected that, as the driver continues to grow, various other abstractions
-will be needed.
+## Architecture
 
+![Nova Architecture with vGPUs](./nova-core-vm.png)
 
-## Utilized Common Rust Infrastructure
+The overall GPU driver is split into two parts:
 
-Nova depends on the Rust for Linux `staging/*` [branches](Branches.md).
+1. "Nova-Core", living in `drivers/gpu/nova-core/`. Nova-Core implements
+   the fundamental interaction with the hardware (through PCI etc.) and,
+   notably, boots up the GSP and interacts with it through a command queue.
+2. "Nova-DRM" (the official name is actually just "Nova", but to avoid
+   confusion developers usually call it "Nova-DRM"), living in
+   `drivers/gpu/drm/nova/`. This is the actual graphics driver,
+   implementing all the typical DRM interfaces for userspace.
 
+This split architecture allows for virtualizing GPUs: Nova-Core can be used to
+instruct the GPU's firmware to spawn a new PCI virtual function (Through
+[SR-IOV](https://docs.kernel.org/PCI/pci-iov-howto.html)). This virtual
+function can be passed to a virtual machine, which, for example, can run
+another Linux with another Nova-Core bound to the virtual GPU. On top, Nova-DRM
+can be utilized as a conventional GPU driver to use the vGPU.
 
-## Contributing
+A main advantage of this design is that the amount of software running on the
+host (where crashes would be far more fatal than inside of a VM) is kept small,
+which contributes to stability.
 
-As with every real open source program, help and participation is highly welcome!
+Of course, it is also possible to use Nova-Core + Nova-DRM on one physical
+machine (not shown in the graphics), then directly using the GPU through Mesa
```

Comment:
I'd write this as:

> It is also possible to use Nova-Core + Nova-DRM on one physical machine, in order to expose a DRM compatible uAPI to the host userspace.

I think the "of course" and the hint that the diagram shows something different is unnecessary.

For the second part of the sentence, it's not necessarily bound to Mesa, anyone could implement against Nova's uAPI.